src/tamm/proc_group.hpp

```diff
@@ -375,13 +375,36 @@ class ProcGroup {
 
   template<typename T>
   void allgather(const T* sbuf, T* rbuf) {
-    throw std::runtime_error("allgather unsupported");
+    EXPECTS(sbuf != nullptr);
+
+    auto             nranks = size().value();
+    upcxx::promise<> p(nranks);
+
+    for(int r = 0; r < nranks; ++r)
+      upcxx::broadcast(*sbuf, r, *team()).then([&rbuf, &p, r](T result) {
+        rbuf[r] = result;
+        p.fulfill_anonymous(1);
+      });
+
+    p.get_future().wait();
   }
 
   template<typename T>
   void allgather(const T* sbuf, int scount, T* rbuf, int rcount) {
-    throw std::runtime_error("allgather unsupported");
+    EXPECTS(sbuf != nullptr);
+
```

Comment:
I don't understand why this function (or the MPI flavor) take separate arguments of `rcount` and `scount`. These counts can only meaningfully differ if the send and recv data types differ, but they are both hard-coded to `T`, so passing unequal values would lead to Very Bad Things.

```suggestion
    EXPECTS(rcount == scount);
```